stdlib/public/runtime/Paths.cpp

```diff
@@ -0,0 +1,196 @@
+swift_getRuntimePath()
+{
+  swift::once(runtimePathToken, _swift_initRuntimePath, nullptr);
+  return runtimePath;
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRootPath()
+{
+  swift::once(rootPathToken, _swift_initRootPath, nullptr);
+  return rootPath;
+}
+
+namespace {
+
+void
+_swift_initRootPath(void *)
+{
+  // SWIFT_ROOT overrides the path returned by this function
+  char *swiftRoot = getenv("SWIFT_ROOT");
+  if (swiftRoot) {
+    size_t len = std::strlen(swiftRoot);
+    if (len) {
+      // Make sure it has a trailing path separator
+      if (_swift_isPathSep(swiftRoot[len - 1]))
+        rootPath = ::strdup(swiftRoot);
+      else {
+        char *path = (char *)std::malloc(len + 2);
+        std::memcpy(path, swiftRoot, len);
+        std::memcpy(path + len, PATHSEP_STR, 2);
+        rootPath = path;
+      }
+      return;
+    }
+  }
+
+  // Otherwise, get the runtime path and remove /lib/libswiftCore.(dylib|dll|so)
+  const char *runtimePath = swift_getRuntimePath();
+  const char *end = runtimePath + std::strlen(runtimePath);
+  const char *ptr = end;
+  int count = 0;
+
+  while (ptr > runtimePath) {
+    int ch = *--ptr;
+    if (_swift_isPathSep(ch)) {
+      if (++count == 2)
+        break;
+    }
+  }
+
+  if (count != 2) {
+    swift::fatalError(/* flags = */ 0,
+                      "libswiftCore should be installed in a `lib` "
+                      "directory: %s\n",
+                      runtimePath);
+  }
+
+  // Copy the path including the trailing '/'
+  size_t len = ptr - runtimePath + 1;
+  char *path = (char *)std::malloc(len + 1);
+  std::memcpy(path, runtimePath, len);
+  path[len] = 0;
+
+  rootPath = path;
+}
+
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getAuxiliaryExecutablePath(const char *name)
+{
+  const char *rootPath = swift_getRootPath();
+
+  // Allocate a buffer for the result
+  const size_t nameLen = std::strlen(name);
+  const size_t extLen = std::strlen(EXE_EXT);
+  const size_t rootPathLen = std::strlen(rootPath);
+  const size_t libexecLen = 8;
+  char *result = (char *)std::malloc(rootPathLen + libexecLen
+                                     + nameLen + extLen + 1);
+  char *ptr = result;
+
+  std::memcpy(ptr, rootPath, rootPathLen); ptr += rootPathLen;
+  std::memcpy(ptr, "libexec" PATHSEP_STR, libexecLen); ptr += libexecLen;
+  std::memcpy(ptr, name, nameLen); ptr += nameLen;
+  std::memcpy(ptr, EXE_EXT, extLen + 1);
+
+  return result;
+}
+
+namespace {
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+void
+_swift_initRuntimePath(void *) {
+  Dl_info dli;
+
+  int ret = ::dladdr((void *)_swift_initRuntimePath, &dli);
+
+  if (!ret) {
+    swift::fatalError(/* flags = */ 0,
+                      "Unable to obtain Swift runtime path\n");
+  }
+
+  runtimePath = ::strdup(dli.dli_fname);
+}
+#else
+void
+_swift_initRuntimePath(void *) {
+  const DWORD dwBufSize = 4096;
+  LPSTR lpFilename = (LPSTR)std::malloc(bufsize);
+
+  DWORD dwRet = GetMappedFileNameA(GetCurrentProcess(),
```

Comment:
We really need to use `GetMappedFileNameW`.  However, that is not going to be a realpath.  We should use `GetFinalPathNameByHandleW` taking the module handle to map it to the realpath.  Ascii paths may be invalid if the codepoint cannot be mapped into the ASCII plane.

Reply:
I did wonder about that, and whether or not that's genuinely going to be a problem.

That then begs a couple more questions:

1. Do we want this to return UTF-8 on Windows. It could do that, and that would play nice with Swift if anyone wanted to call these functions from Swift code. _However_ it also wouldn't work with the ANSI Windows APIs (including the C library, right?) unless someone has forced the code page to UTF-8.
2. Alternatively, we could have it return a wide string on Windows — but that presumably won't play so nicely with Swift code, and the API will then not match that on other platforms.

Reply:
I think that the right thing to do is what we do elsewhere: system uses UTF-16 unconditionally.  Going from system to Swift, we convert and convert again from Swift to system.

Reply:
My only concern here is that calling `open()` from Swift on Windows with a path derived from one of these functions is then not guaranteed to work without setting the code page to UTF-8. I suppose that's a general hazard on Windows, but it mightn't be obvious to the uninitiated.

Reply:
Hmmm. I don't think `GetFinalPathNameByHandleW()` takes an `HMODULE`. In fact, I'm fairly sure it won't (it's been a while so I might be misremembering, but an `HMODULE` is really a pointer to the `IMAGE_DOS_HEADER`, IIRC, rather than a file handle which is what `GetFinalPathNameByHandleW()` accepts).

We _could_ use `GetModuleFileNameExW()` rather than `GetMappedFileNameW()`, but that means we'd need the `HMODULE` for `libswiftCore.dll` — admittedly that's rather easy to get (via `__ImageBase`), but I'm not sure we gain that much in practice.

Reply:
Or are you suggesting that we should use `CreateFile()` to open the path we find, and pass that through `GetFinalPathNameByHandleW()`? I can see _an_ argument for doing something like that, and similarly calling `realpath()` on UNIX, in that it would mean you could symlink just `libswiftCore.dylib` and then resolve auxiliary items relative to the _target_ of the symlink. But equally, that would mean that you _couldn't_ symlink it and then resolve items relative to the location of the symbolic link, which is potentially equally useful (e.g. you could then replace just the auxiliary data).

Reply:
(On the `GetModuleFileNameExW()` versus `GetMappedFileNameW()` front, I note that the docs for the former claim it may fail if the module list is not yet initialised or if it changes while the function is being called. There is no such note for the latter, so it might in fact be more robust.)

Reply:
I now understand why you thought `GetFinalPathNameByHandleW()` would be needed. It's not _quite_ the right function though, and to use it we'd have to use `NtCreateFile()` which I'm reluctant to do because that's an internal API. So I've solved it in a similar way to the code MS gives for mapping these filenames on previous Windows versions.